### PR TITLE
adds and uses new method to insert objects and prefixes into stat cache

### PR DESCRIPTION
**Pull request by ankitaluthra1**
### Description
adds and uses new method to insert objects and prefixes into stat cache.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: internal/storage/caching/fast_stat_bucket.go [hunk 1]
```diff
@@ -83,6 +84,32 @@ func (b *fastStatBucket) insertMultiple(objs []*gcs.Object) {
...
+	for _, o := range listing.Objects {
+		if !strings.HasSuffix(o.Name, "/") {
+			m := storageutil.ConvertObjToMinObject(o)
+			b.cache.Insert(m, expiration)
+		}
+	}
+
+	for _, p := range listing.CollapsedRuns {
+		if !strings.HasSuffix(p, "/") {
+			_ = fmt.Errorf("error in prefix name: %s", p)
```
[the remark this answers is not in the document]
Comment (ankitaluthra1) [reply internal/storage/caching/fast_stat_bucket.go]: Ideally this case should not happen, but ee should at least log it as error as in case anomaly, we should be able to debug.

File: internal/fs/inode/dir.go [hunk 2]
```diff
@@ -691,10 +691,10 @@ func (d *dirInode) readObjects(
 		}
 	}
 
-	// Return an appropriate continuation token, if any.
+	// return an appropriate continuation token, if any.
 	newTok = listing.ContinuationToken
 
-	if !d.implicitDirs {
+	if !d.implicitDirs && !d.isHNSEnabled {
```
Comment (Tulsishah) [inline internal/fs/inode/dir.go:697]: Can you please use `!d.isBucketHierarchical() `

File: internal/storage/caching/fast_stat_bucket.go [hunk 3]
```diff
@@ -83,6 +85,36 @@ func (b *fastStatBucket) insertMultiple(objs []*gcs.Object) {
 	}
 }
 
+func (b *fastStatBucket) insertListing(listing *gcs.Listing) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	expiration := b.clock.Now().Add(b.ttl)
+
+	// save the objects in cache
```
Comment (Tulsishah) [inline internal/storage/caching/fast_stat_bucket.go:94]: nit: Start with capital later and end with full stop.

**Review by Tulsishah — DISMISSED** [review]
Can you please validate manually also, if it is working properly?

**Review by charith87 — DISMISSED** [review]
Approving go/lgtm-with-comments

File: internal/fs/inode/dir.go (hunk 2 above)
Comment (ankitaluthra1) [reply internal/fs/inode/dir.go:697]: yes actually d.isBucketHierarchical() suffices both.. Tulsi added it later so I think its redundant at older places and places where I updated the code... yes agreed we can try and remove as when we touch the check code again in any new changes

File: internal/storage/caching/fast_stat_bucket.go [hunk 4]
```diff
@@ -253,9 +285,13 @@ func (b *fastStatBucket) ListObjects(
 		return
 	}
 
-	// Note anything we found.
-	b.insertMultiple(listing.Objects)
+	if b.BucketType() == gcs.Hierarchical {
+		b.insertListing(listing)
```
[the remark this answers is not in the document]
Comment (ankitaluthra1) [reply internal/storage/caching/fast_stat_bucket.go:289]: yes done

File: internal/fs/inode/dir.go (hunk 2 above)
Comment (charith87) [reply internal/fs/inode/dir.go]: Okk. Pls add a todo in code as a comment otherwise its easy to miss them if you are not doing it now.

File: internal/storage/caching/fast_stat_bucket.go (hunk 3 above)
Comment (ankitaluthra1) [reply internal/storage/caching/fast_stat_bucket.go:94]: done

**Review by charith87 — APPROVED** [review]

File: internal/storage/caching/fast_stat_bucket.go [hunk 5]
```diff
@@ -83,6 +85,36 @@ func (b *fastStatBucket) insertMultiple(objs []*gcs.Object) {
...
+		if !strings.HasSuffix(o.Name, "/") {
+			m := storageutil.ConvertObjToMinObject(o)
+			b.cache.Insert(m, expiration)
+		}
+	}
+
+	for _, p := range listing.CollapsedRuns {
+		if !strings.HasSuffix(p, "/") {
+			// log the error for incorrect prefix but don't fail the operation
+			logger.Errorf("error in prefix name: %s", p)
```
Comment (gargnitingoogle) [inline internal/storage/caching/fast_stat_bucket.go:107]: Should this be a warning instead of an error? Asking this because this is a GCS-caused issue, and it does not affect user in any way.

File: internal/storage/caching/fast_stat_bucket.go [hunk 6]
```diff
@@ -83,6 +85,36 @@ func (b *fastStatBucket) insertMultiple(objs []*gcs.Object) {
...
+		} else {
+			f := &gcs.Folder{
+				Name: p,
+			}
+			b.cache.InsertFolder(f, expiration)
+		}
+	}
+
```
Comment (gargnitingoogle) [inline internal/storage/caching/fast_stat_bucket.go:115]: remove unnecessary empty line.

File: internal/storage/caching/fast_stat_bucket.go (hunk 5 above)
Comment (gargnitingoogle) [reply internal/storage/caching/fast_stat_bucket.go:107]: Anyway this error/warning must have been also logged at the `bucketHandle` level, so it's just piling on error message after error message for the same issue.

File: internal/storage/caching/fast_stat_bucket.go [hunk 7]
```diff
@@ -253,9 +285,13 @@ func (b *fastStatBucket) ListObjects(
...
 	}
 
-	// Note anything we found.
-	b.insertMultiple(listing.Objects)
+	if b.BucketType() == gcs.Hierarchical {
+		b.insertHierarchicalListing(listing)
+		return
+	}
 
+	// note anything we found.
```
Comment (gargnitingoogle) [inline internal/storage/caching/fast_stat_bucket.go:293]: Please fix the comment. Original comment was good already.

File: tools/integration_tests/util/operations/file_operations.go [hunk 8]
```diff
@@ -41,7 +41,7 @@ const (
 	// TimeSlop The radius we use for "expect mtime is within"-style assertions as kernel
 	// time can be slightly out of sync of time.Now().
 	// Ref: https://github.com/golang/go/issues/33510
-	TimeSlop = 25 * time.Millisecond
+	TimeSlop = 100 * time.Millisecond
```
Comment (gargnitingoogle) [inline tools/integration_tests/util/operations/file_operations.go:44]: Why is this change needed? I read the link but couldn't make sense why this needs to be increased here.

**Review by gargnitingoogle — COMMENTED** [review]
Added some minor comments

**Review by charith87 — APPROVED** [review]

File: internal/storage/caching/fast_stat_bucket.go (hunk 5 above)
Comment (ankitaluthra1) [reply internal/storage/caching/fast_stat_bucket.go:107]: Anyway this error/warning must have been also logged at the bucketHandle level, so it's just piling on error message after error message for the same issue. > since this error is gulped here and not returned it will not be logged at bucket handle level, 
added error level here as this is an error only, should be logged at log level info also

File: internal/storage/caching/fast_stat_bucket.go (hunk 7 above)
Comment (ankitaluthra1) [reply internal/storage/caching/fast_stat_bucket.go:293]: noted, will update in subsequent PRs as dont want to block this change on it.
